crates/wasmtime/src/runtime/component/concurrent/futures_and_streams.rs

```diff
@@ -1,14 +1,404 @@
+        }
+    }
+}
+
+unsafe impl<T> func::Lower for StreamReader<T> {
+    fn lower<U>(
+        &self,
+        cx: &mut LowerContext<'_, U>,
+        ty: InterfaceType,
+        dst: &mut MaybeUninit<Self::Lower>,
+    ) -> Result<()> {
+        self.lower_to_index(cx, ty)?
+            .lower(cx, InterfaceType::U32, dst)
+    }
+
+    fn store<U>(
+        &self,
+        cx: &mut LowerContext<'_, U>,
+        ty: InterfaceType,
+        offset: usize,
+    ) -> Result<()> {
+        self.lower_to_index(cx, ty)?
+            .store(cx, InterfaceType::U32, offset)
+    }
+}
+
+unsafe impl<T> func::Lift for StreamReader<T> {
+    fn lift(cx: &mut LiftContext<'_>, ty: InterfaceType, src: &Self::Lower) -> Result<Self> {
+        let index = u32::lift(cx, InterfaceType::U32, src)?;
+        Self::lift_from_index(cx, ty, index)
+    }
+
+    fn load(cx: &mut LiftContext<'_>, ty: InterfaceType, bytes: &[u8]) -> Result<Self> {
+        let index = u32::load(cx, InterfaceType::U32, bytes)?;
+        Self::lift_from_index(cx, ty, index)
+    }
+}
+
 /// Represents a Component Model `error-context`.
-pub struct ErrorContext {}
+pub struct ErrorContext {
+    rep: u32,
+}
+
+impl ErrorContext {
+    pub(crate) fn new(rep: u32) -> Self {
+        Self { rep }
+    }
+
+    /// Convert this `ErrorContext` into a [`Val`].
+    pub fn into_val(self) -> Val {
+        Val::ErrorContext(ErrorContextAny(self.rep))
+    }
+
+    fn lower_to_index<U>(&self, cx: &mut LowerContext<'_, U>, ty: InterfaceType) -> Result<u32> {
+        match ty {
+            InterfaceType::ErrorContext(dst) => {
+                let tbl = unsafe {
+                    &mut (*cx.instance)
+                        .component_error_context_tables()
+                        .get_mut(dst)
+                        .expect("error context table index present in (sub)component table during lower")
+                };
+
+                if let Some((dst_idx, dst_state)) = tbl.get_mut_by_rep(self.rep) {
+                    dst_state.0 += 1;
+                    Ok(dst_idx)
+                } else {
+                    tbl.insert(self.rep, LocalErrorContextRefCount(1))
+                }
+            }
+            _ => func::bad_type_info(),
+        }
+    }
+
+    fn lift_from_index(cx: &mut LiftContext<'_>, ty: InterfaceType, index: u32) -> Result<Self> {
+        match ty {
+            InterfaceType::ErrorContext(src) => {
+                let (rep, _) = unsafe {
+                    (*cx.instance)
+                        .component_error_context_tables()
+                        .get_mut(src)
+                        .expect(
+                            "error context table index present in (sub)component table during lift",
+                        )
+                        .get_mut_by_index(index)?
+                };
+
+                Ok(Self { rep })
+            }
+            _ => func::bad_type_info(),
+        }
+    }
```

Comment:
I'll definitely admit that I'm very lost when it comes to error-context here. I see a local refcount table and a global refcount table but I don't actually see the global refcount table being used. I'm also not really sure what `rep` is in the context of an error-context so I'm pretty lost here.

While comments would help here is this something that may be best deferred to a future PR? For example ideally there'd be a test exercising this lifting and lowering and testing various errors here and things like that, but I suspect such a test isn't quite ready yet?

Reply:
Yeah, I can replace this stuff with stubs.  My approach with this PR was to run the lift.wast test, hit a `todo!()`, replace the `todo!()` with code copied from #9582, let the compiler tell me what _other_ code _that_ code needed, then either copying more code or adding stubs to make the compiler happy.  Looks like I copied over too much, and that's making the review more difficult; sorry about that.  I'll trim it down.